### PR TITLE
*: unified the cancel/kill logic in tidb

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/47075

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: util/sqlkiller/sqlkiller.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	connid uint64
+}
+
+// NewSQLKiller creates a new SqlKiller.
+func NewSQLKiller(connid uint64, status *uint32) *SQLKiller {
+	return &SQLKiller{connid: connid, status: status}
+}
+
+// SendKillSignal sends a kill signal to the query.
+func (killer *SQLKiller) SendKillSignal(reason uint32) {
```
Comment (yibin87) [inline util/sqlkiller/sqlkiller.go:46]: might return bool value to indicate if send successful?

Comment (wshwsh12) [reply util/sqlkiller/sqlkiller.go:46]: I don't think it's necessary to return if send was successful. Only when a kill signal succeeds in sending, other kill signals may fail to send. In this case, the SQL is killed regardless of whether subsequent signals are successfully sent.

**Review by YangKeao — COMMENTED** [review]
Rest LGTM

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by XuHuaiyu — DISMISSED** [review]

File: pkg/util/sqlkiller/sqlkiller.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	UnspecifiedKillSignal killSignal = iota
+	QueryInterrupted
+	MaxExecTimeExceeded
+	QueryMemoryExceeded
+	ServerMemoryExceeded
+)
+
+// SQLKiller is used to kill a query.
+type SQLKiller struct {
+	Status killSignal
```
Comment (XuHuaiyu) [inline pkg/util/sqlkiller/sqlkiller.go:38]:
```suggestion
	Signal killSignal
```

File: pkg/util/sqlkiller/sqlkiller.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/pkg/util/dbterror/exeerrors"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+type killSignal = uint32
+
+// Kill reasons.
```
Comment (XuHuaiyu) [inline pkg/util/sqlkiller/sqlkiller.go:27]:
```suggestion
```

File: pkg/util/memory/action.go [hunk 4]
```diff
@@ -183,22 +168,17 @@ func (a *PanicOnExceed) Action(t *Tracker) {
...
 }
 
-// SetInvoker sets the invoker of the Action.
-func (a *PanicOnExceed) SetInvoker(invoker ActionInvoker) {
-	a.invoker = invoker
-}
-
 var (
 	errMemExceedThreshold = dbterror.ClassUtil.NewStd(errno.ErrMemExceedThreshold)
 )
```
Comment (XuHuaiyu) [inline pkg/util/memory/action.go:184]: line 207~212 is only used in test, we can remove them

File: pkg/util/sqlkiller/sqlkiller.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	case QueryMemoryExceeded:
+		return exeerrors.ErrMemoryExceed.GenWithStackByArgs("a single SQL query", "tidb_mem_quota_query limit", killer.ConnID)
+	case ServerMemoryExceeded:
+		logutil.BgLogger().Warn("global memory controller, NeedKill signal is received successfully",
+			zap.Uint64("conn", killer.ConnID))
+		return exeerrors.ErrMemoryExceed.GenWithStackByArgs("the tidb-server instance and this query is currently using the most memory", "tidb_server_memory_limit", killer.ConnID)
+	}
+	return nil
+}
+
```
Comment (XuHuaiyu) [inline pkg/util/sqlkiller/sqlkiller.go:64]:
``` go
var (
    errQueryMemoryeExceed = exeerrors.ErrMemoryExceed.GenWithStackByArgs("a single SQL query", "tidb_mem_quota_query limit", killer.ConnID)
    errServerMemoryExceeded = ...
)
```

File: pkg/executor/executor_failpoint_test.go [hunk 6]
```diff
@@ -246,18 +246,18 @@ func TestKillTableReader(t *testing.T) {
 	tk.MustExec("create table t (a int)")
 	tk.MustExec("insert into t values (1),(2),(3)")
 	tk.MustExec("set @@tidb_distsql_scan_concurrency=1")
-	atomic.StoreUint32(&tk.Session().GetSessionVars().Killed, 0)
+	atomic.StoreUint32(&tk.Session().GetSessionVars().SQLKiller.Signal, 0)
```
Comment (XuHuaiyu) [inline pkg/executor/executor_failpoint_test.go:249]: use Reset here

File: pkg/executor/executor_failpoint_test.go [hunk 7]
```diff
@@ -246,18 +246,18 @@ func TestKillTableReader(t *testing.T) {
...
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		time.Sleep(1 * time.Second)
-		err := tk.QueryToErr("select * from t")
-		require.Error(t, err)
-		require.Equal(t, int(exeerrors.ErrQueryInterrupted.Code()), int(terror.ToSQLError(errors.Cause(err).(*terror.Error)).Code))
+		time.Sleep(300 * time.Millisecond)
+		atomic.StoreUint32(&tk.Session().GetSessionVars().SQLKiller.Signal, 1)
```
Comment (XuHuaiyu) [inline pkg/executor/executor_failpoint_test.go:256]: Use SendSignal here

File: pkg/executor/test/tiflashtest/tiflash_test.go [hunk 8]
```diff
@@ -682,7 +681,7 @@ func TestCancelMppTasks(t *testing.T) {
 	// mock executor does not support use outer table as build side for outer join, so need to
 	// force the inner table as build side
 	tk.MustExec("set tidb_opt_mpp_outer_join_fixed_build_side=1")
-	atomic.StoreUint32(&tk.Session().GetSessionVars().Killed, 0)
+	atomic.StoreUint32(&tk.Session().GetSessionVars().SQLKiller.Signal, 0)
```
Comment (XuHuaiyu) [inline pkg/executor/test/tiflashtest/tiflash_test.go:684]: killer.Reset

File: pkg/executor/test/tiflashtest/tiflash_test.go [hunk 9]
```diff
@@ -693,7 +692,7 @@ func TestCancelMppTasks(t *testing.T) {
 		require.Equal(t, int(exeerrors.ErrQueryInterrupted.Code()), int(terror.ToSQLError(errors.Cause(err).(*terror.Error)).Code))
 	}()
 	time.Sleep(1 * time.Second)
-	atomic.StoreUint32(&tk.Session().GetSessionVars().Killed, 1)
+	atomic.StoreUint32(&tk.Session().GetSessionVars().SQLKiller.Signal, 1)
```
Comment (XuHuaiyu) [inline pkg/executor/test/tiflashtest/tiflash_test.go:695]: Killer.SendSignal

File: pkg/executor/test/unstabletest/unstable_test.go [hunk 10]
```diff
@@ -81,19 +82,19 @@ func TestGlobalMemoryControl(t *testing.T) {
 	time.Sleep(500 * time.Millisecond) // The check goroutine checks the memory usage every 100ms. The Sleep() make sure that Top1Tracker can be Canceled.
 
 	// Kill Top1
-	require.False(t, tracker1.NeedKill.Load())
-	require.False(t, tracker2.NeedKill.Load())
-	require.True(t, tracker3.NeedKill.Load())
+	require.False(t, atomic.LoadUint32(&tracker1.Killer.Signal) > 0)
```
Comment (XuHuaiyu) [inline pkg/executor/test/unstabletest/unstable_test.go:85]: Use HandleSignal and check the error code?

File: pkg/executor/test/unstabletest/unstable_test.go [hunk 11]
```diff
@@ -81,19 +82,19 @@ func TestGlobalMemoryControl(t *testing.T) {
...
 		func() {
 			tracker3.Consume(1)
 		}, func(r interface{}) {
-			require.True(t, strings.Contains(r.(string), memory.PanicMemoryExceedWarnMsg+memory.WarnMsgSuffixForInstance))
+			require.True(t, exeerrors.ErrMemoryExceedForInstance.Equal(r.(error)))
 		})
 	tracker2.Consume(300 << 20) // Sum 500MB, Not Panic, Waiting t3 cancel finish.
 	time.Sleep(500 * time.Millisecond)
-	require.False(t, tracker2.NeedKill.Load())
+	require.False(t, atomic.LoadUint32(&tracker2.Killer.Signal) > 0)
```
Comment (XuHuaiyu) [inline pkg/executor/test/unstabletest/unstable_test.go:97]: ditto

File: pkg/expression/builtin_miscellaneous_vec_test.go [hunk 12]
```diff
@@ -215,7 +215,7 @@ func TestSleepVectorized(t *testing.T) {
 	start = time.Now()
 	go func() {
 		time.Sleep(1 * time.Second)
-		atomic.CompareAndSwapUint32(&ctx.GetSessionVars().Killed, 0, 1)
+		atomic.CompareAndSwapUint32(&ctx.GetSessionVars().SQLKiller.Signal, 0, 1)
```
Comment (XuHuaiyu) [inline pkg/expression/builtin_miscellaneous_vec_test.go:218]: SetSignal

File: pkg/expression/evaluator_test.go [hunk 13]
```diff
@@ -152,7 +152,7 @@ func TestSleep(t *testing.T) {
 	start = time.Now()
 	go func() {
 		time.Sleep(1 * time.Second)
-		atomic.CompareAndSwapUint32(&ctx.GetSessionVars().Killed, 0, 1)
+		atomic.CompareAndSwapUint32(&ctx.GetSessionVars().SQLKiller.Signal, 0, 1)
```
Comment (XuHuaiyu) [inline pkg/expression/evaluator_test.go:155]: SendSignal

File: pkg/server/conn_test.go [hunk 14]
```diff
@@ -720,22 +720,22 @@ func TestConnExecutionTimeout(t *testing.T) {
 	require.Equal(t, "[executor:3024]Query execution was interrupted, maximum statement execution time exceeded", err.Error())
 
 	// Killed because of max execution time, reset Killed to 0.
-	atomic.CompareAndSwapUint32(&tk.Session().GetSessionVars().Killed, 2, 0)
+	atomic.CompareAndSwapUint32(&tk.Session().GetSessionVars().SQLKiller.Signal, 2, 0)
```
Comment (XuHuaiyu) [inline pkg/server/conn_test.go:723]: SendSignal

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 15]
```diff
@@ -728,7 +728,7 @@ func TestWaitLockKill(t *testing.T) {
 	go func() {
 		time.Sleep(500 * time.Millisecond)
 		sessVars := tk2.Session().GetSessionVars()
-		succ := atomic.CompareAndSwapUint32(&sessVars.Killed, 0, 1)
+		succ := atomic.CompareAndSwapUint32(&sessVars.SQLKiller.Signal, 0, 1)
```
Comment (XuHuaiyu) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:731]: SendSignal

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 16]
```diff
@@ -756,12 +756,12 @@ func TestKillStopTTLManager(t *testing.T) {
 	tk2.MustExec("begin pessimistic")
 	tk.MustQuery("select * from test_kill where id = 1 for update")
 	sessVars := tk.Session().GetSessionVars()
-	succ := atomic.CompareAndSwapUint32(&sessVars.Killed, 0, 1)
+	succ := atomic.CompareAndSwapUint32(&sessVars.SQLKiller.Signal, 0, 1)
```
Comment (XuHuaiyu) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:759]: ditto

**Review by XuHuaiyu — COMMENTED** [review]
We'd better use Send/HandleSignal in the test

File: pkg/executor/distsql.go [hunk 17]
```diff
@@ -1168,7 +1168,11 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
 	defer func() {
 		if r := recover(); r != nil {
 			logutil.Logger(ctx).Error("tableWorker in IndexLookUpExecutor panicked", zap.Any("recover", r), zap.Stack("stack"))
-			task.doneCh <- errors.Errorf("%v", r)
+			err := errors.Errorf("%v", r)
```
Comment (XuHuaiyu) [inline pkg/executor/distsql.go:1171]: why make this modification?

Comment (wshwsh12) [reply pkg/executor/distsql.go:1171]: *error.fundamental is not a *error.Error, so we can't keep the error code, or check the error is the same type...

File: pkg/executor/aggregate/agg_util.go [hunk 18]
```diff
@@ -59,8 +59,8 @@ func closeBaseExecutor(b *exec.BaseExecutor) {
 }
 
 func recoveryHashAgg(output chan *AfFinalResult, r interface{}) {
-	err := errors.Errorf("%v", r)
-	output <- &AfFinalResult{err: errors.Errorf("%v", r)}
+	err := util.GetRecoverError(r)
+	output <- &AfFinalResult{err: util.GetRecoverError(r)}
```
Comment (XuHuaiyu) [inline pkg/executor/aggregate/agg_util.go:63]:
```suggestion
	output <- &AfFinalResult{err: err}
```

Comment (wshwsh12) [reply pkg/executor/aggregate/agg_util.go]: done

**Review by XuHuaiyu — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by easonn7** [comment]
/approve

There is no sysvar change.

**Pull request merged**